src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1636 @@
+  // behavior we need there.
+  std::unordered_map<Expression*, Expression*> childParents;
+};
+
+// Walk the wasm and find all the links we need to care about, and the locations
+// and roots related to them. This builds up a CollectedFuncInfo data structure.
+// After all InfoCollectors run, those data structures will be merged and the
+// main flow will begin.
+struct InfoCollector
+  : public PostWalker<InfoCollector, OverriddenVisitor<InfoCollector>> {
+  CollectedFuncInfo& info;
+
+  InfoCollector(CollectedFuncInfo& info) : info(info) {}
+
+  // Check if a type is relevant for us. If not, we can ignore it entirely.
+  bool isRelevant(Type type) {
+    if (type == Type::unreachable || type == Type::none) {
+      return false;
+    }
+    if (type.isTuple()) {
+      for (auto t : type) {
+        if (isRelevant(t)) {
+          return true;
+        }
+      }
+    }
+    if (type.isRef() && getTypeSystem() != TypeSystem::Nominal &&
+        getTypeSystem() != TypeSystem::Isorecursive) {
+      // We need explicit supers in the SubTyping helper class. Without that,
+      // cannot handle refs, and consider them irrelevant.
+      return false;
+    }
+    return true;
+  }
+
+  bool isRelevant(Signature sig) {
+    return isRelevant(sig.params) || isRelevant(sig.results);
+  }
+
+  bool isRelevant(Expression* curr) { return curr && isRelevant(curr->type); }
+
+  template<typename T> bool isRelevant(const T& vec) {
+    for (auto* expr : vec) {
+      if (isRelevant(expr->type)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Each visit*() call is responsible for connecting the children of a node to
+  // that node. Responsibility for connecting the node's output to anywhere
+  // else (another expression or the function itself, if we are at the top
+  // level) is the responsibility of the outside.
+
+  void visitBlock(Block* curr) {
+    if (curr->list.empty()) {
+      return;
+    }
+
+    // Values sent to breaks to this block must be received here.
+    handleBreakTarget(curr);
+
+    // The final item in the block can flow a value to here as well.
+    receiveChildValue(curr->list.back(), curr);
+  }
+  void visitIf(If* curr) {
+    // Each arm may flow out a value.
+    receiveChildValue(curr->ifTrue, curr);
+    receiveChildValue(curr->ifFalse, curr);
+  }
+  void visitLoop(Loop* curr) { receiveChildValue(curr->body, curr); }
+  void visitBreak(Break* curr) {
+    // Connect the value (if present) to the break target.
+    handleBreakValue(curr);
+
+    // The value may also flow through in a br_if (the type will indicate that,
+    // which receiveChildValue will notice).
+    receiveChildValue(curr->value, curr);
+  }
+  void visitSwitch(Switch* curr) { handleBreakValue(curr); }
+  void visitLoad(Load* curr) {
+    // We could infer the exact type here, but as no subtyping is possible, it
+    // would have no benefit, so just add a generic root (which will be "Many").
+    addRoot(curr);
```

Comment:
Can't this be an `ExactType` too?

Reply:
The meeting today made this clearer. I think this can be summarized along the lines of "we don't need to send redundant information that can be supplied by the expression types."